File: tests/fixtures/cache/cache_fixture_helpers.R
statgen_repo_root <- function() {
  normalizePath(file.path(statgen_cache_dir, "..", "..", ".."), mustWork = TRUE)
}

statgen_source_r_package <- function(repo_root) {
  r_files <- list.files(file.path(repo_root, "R-package", "R"), pattern = "[.]R$", full.names = TRUE)
  for (path in r_files) {
    source(path)
  }
  set_verbosity("quiet")
}

statgen_cache_fixture_reference <- function(repo_root) {
  load_reference(file.path("tests", "fixtures", "reference", "sharded", "@.bim"))
}

statgen_cache_fixture_annotations <- function(repo_root, reference) {
  fixture_dir <- file.path("tests", "fixtures")
  binary <- load_annotations(
    file.path(fixture_dir, "annotations", c("anno1.bed", "anno2.bed")),
    reference
  )
  continuous <- load_annotation(
    file.path(fixture_dir, "annotations", "continuous.annot"),
    reference,
    header = TRUE,
    value_columns = c("score", "weight"),
    annotation_metadata_path = file.path(fixture_dir, "annotations", "continuous.meta")
  )
  union_annotations(binary, continuous)
}

statgen_generate_r_cache_fixture <- function(object_name) {
  repo_root <- statgen_repo_root()
  cache_dir <- file.path(repo_root, "tests", "fixtures", "cache")
  dir.create(cache_dir, recursive = TRUE, showWarnings = FALSE)
  statgen_source_r_package(repo_root)
  reference <- statgen_cache_fixture_reference(repo_root)
  fixture_dir <- file.path("tests", "fixtures")

  if (identical(object_name, "reference")) {
    save_reference_cache(reference, file.path(cache_dir, "reference_r_reference_cache_0_1.rds"))
  } else if (identical(object_name, "sumstats")) {
    panel <- load_sumstats(file.path(fixture_dir, "sumstats", "traits_complete.tsv.gz"), reference)
    save_sumstats_cache(panel, file.path(cache_dir, "sumstats_r_sumstats_cache_0_1.rds"))
  } else if (identical(object_name, "annotations")) {
    save_annotations_cache(
      statgen_cache_fixture_annotations(repo_root, reference),
      file.path(cache_dir, "annotations_r_annotations_cache_0_2.rds")
    )
  } else if (identical(object_name, "genotype")) {
    panel <- load_genotype(file.path("tests", "fixtures", "genotype", "sharded", "@"), reference)
    save_genotype_cache(panel, file.path(cache_dir, "genotype_r_genotype_cache_0_1.rds"))
  } else {
    stop(sprintf("unknown cache fixture object: %s", object_name), call. = FALSE)
  }
}

File: tests/fixtures/cache/cache_fixture_helpers.py
#!/usr/bin/env python3
"""Shared helpers for current-schema Python cache fixture generators."""

from pathlib import Path
import sys

CACHE_DIR = Path(__file__).resolve().parent
FIXTURES_DIR = CACHE_DIR.parent
FIXTURES_REL = Path("tests/fixtures")
REPO_ROOT = FIXTURES_DIR.parent.parent
PYTHON_ROOT = REPO_ROOT / "python"
if str(PYTHON_ROOT) not in sys.path:
    sys.path.insert(0, str(PYTHON_ROOT))

from statgen.annotations import load_annotation, load_annotations, save_annotations_cache
from statgen.genotype import load_genotype, save_genotype_cache
from statgen.reference import load_reference, save_reference_cache
from statgen.sumstats import load_sumstats, save_sumstats_cache


def _reference():
    return load_reference(FIXTURES_REL / "reference/sharded/@.bim")


def _annotations(reference):
    binary = load_annotations(
        [
            FIXTURES_REL / "annotations/anno1.bed",
            FIXTURES_REL / "annotations/anno2.bed",
        ],
        reference,
    )
    continuous = load_annotation(
        FIXTURES_REL / "annotations/continuous.annot",
        reference,
        has_header=True,
        value_columns=["score", "weight"],
        annotation_metadata_path=FIXTURES_REL / "annotations/continuous.meta",
    )
    return binary.union_annotations(continuous)


def generate_python_cache_fixture(object_name: str) -> None:
    CACHE_DIR.mkdir(parents=True, exist_ok=True)
    reference = _reference()
    if object_name == "reference":
        save_reference_cache(reference, CACHE_DIR / "reference_python_reference_cache_0_1.npz")
    elif object_name == "sumstats":
        panel = load_sumstats(FIXTURES_REL / "sumstats/traits_complete.tsv.gz", reference)
        save_sumstats_cache(panel, CACHE_DIR / "sumstats_python_sumstats_cache_0_1.npz")
    elif object_name == "annotations":
        save_annotations_cache(_annotations(reference), CACHE_DIR / "annotations_python_annotations_cache_0_2.npz")
    elif object_name == "genotype":
        panel = load_genotype("tests/fixtures/genotype/sharded/@", reference)
        save_genotype_cache(panel, CACHE_DIR / "genotype_python_genotype_cache_0_1.npz")
    else:
        raise ValueError(f"unknown cache fixture object: {object_name}")
